### PR TITLE
Enable hpz based on secondary tensor presence

### PR DESCRIPTION
**Pull request by HeyangQin**
Previously we use a series of forward/backward flags to control if hpz should be enabled on certain allgather call. This PR simplifies this by enabling hpz only when its secondary tensor exists (and invalidating its secondary tensor whenever master weights changes). This should:
1. Prevent potential out-of-sync issue compared with our currently way of overwriting secondary tensor 
2.  Improve throughput because now hpz will be enabled in a lot of different scenarios including i) activation checkpointing, ii) gradient accumulation, iii)`torch.no_grad` context, iv) `model.eval()` mode, v)LoRA frozen weights, vi) gradient overflow

This is to fix https://github.com/microsoft/DeepSpeed/issues/4851

Convergence test:

- llama-2-7b random weights, using https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step1_supervised_finetuning/training_scripts/llama2/run_llama2_7b.sh.

> zero-3 Baseline: Evaluating perplexity, Epoch 4/4: ppl: 5.151907920837402, loss: 1.6393671035766602
> hpz with this PR: ppl: 5.081737518310547, loss: 1.6256532669067383

- llama-2-7b pretrained weights with lora, using https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step1_supervised_finetuning/training_scripts/llama2/run_llama2_7b_lora.sh.

> zero-3 Baseline: Evaluating perplexity, Epoch 4/4: ppl: 1.8326854705810547, loss: 0.6057823896408081
> hpz with this PR: ppl: 1.8326854705810547, loss: 0.6057823896408081

Performance test on 32 V100, still using https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step1_supervised_finetuning/training_scripts/llama2/run_llama2_7b.sh.
- gradient accumulation step = 8
> master branch with hpz: SamplesPerSec=17.567813158654847
> this patch with hpz: SamplesPerSec=24.121657876029225
- lora
> master branch with hpz: SamplesPerSec=33.88883430864484
> this patch with hpz: SamplesPerSec=43.39463460004735

### CODE REVIEW

**Review by samadejacobs — APPROVED** [review]
LGTM, good job @HeyangQin

File: deepspeed/runtime/zero/partitioned_param_coordinator.py [hunk 1]
```diff
@@ -56,14 +56,6 @@ def __setitem__(self, param: Parameter, handle: AllGatherCoalescedHandle) -> Non
 
 
 class PartitionedParameterCoordinator:
-    FORWARD_FETCH_SUBMIT = 'forward_fetch_submit'
```
Comment (tjruwase) [inline deepspeed/runtime/zero/partitioned_param_coordinator.py:59]: Please don't remove profiling code. Instead, please ensure that it is disabled by default.

Comment (HeyangQin) [reply deepspeed/runtime/zero/partitioned_param_coordinator.py:59]: Fixed

Comment (tjruwase) [reply deepspeed/runtime/zero/partitioned_param_coordinator.py:59]: The prefetch and backward profilers (e.g, `FORWARD_PREFETCH_SUBMIT` and `BACKWARD_FETCH_SUBMIT`) are not yet restored.

**Comment by mrwyattii** [comment]
Manually running nightly tests here: https://github.com/microsoft/DeepSpeed/actions/runs/7658819103

File: tests/unit/runtime/zero/test_zeropp.py [hunk 2]
```diff
@@ -92,3 +105,173 @@ def test(self, h_dim: int, n_layers: int, zpg: int) -> None:
...
+            loss = model(batch[0], batch[1])
+            model.backward(loss)
+            model.step()
+
+
+@pytest.mark.nightly
+@pytest.mark.parametrize("model_name", ["gpt2"])
+class TestZeroPPConvergence(DistributedTest):
+    world_size = 4
+    exec_timeout = DistributedTest.exec_timeout * 6  # let's allow 1 hour for this test
```
Comment (mrwyattii) [inline tests/unit/runtime/zero/test_zeropp.py:194]:
```suggestion
```

Test completes in less than 10 minute default timeout, removing this

**Review by mrwyattii — APPROVED** [review]

**Comment by siddharth9820** [comment]
Is hpz safe to use now?
